### PR TITLE
Add Bigarray 'init' functions

### PR DESCRIPTION
**Pull request by yallop**
Add `Bigarray.{Genarray,Array0,Array1,Array2,Array3}.init` functions to match `Array`, `List`, `Float.Array`, `string`, `bytes`:

```ocaml
module Genarray : sig
  (* ... *)

  val init: ('a, 'b) kind -> 'c layout -> int array -> (int array -> 'a) ->
            ('a, 'b, 'c) t
  (** [Genarray.init kind layout dimensions f] returns a new Bigarray [b]
      whose element kind is determined by the parameter [kind] (one of
      [float32], [float64], [int8_signed], etc) and whose layout is
      determined by the parameter [layout] (one of [c_layout] or
      [fortran_layout]).  The [dimensions] parameter is an array of
      integers that indicate the size of the big array in each dimension.
      The length of [dimensions] determines the number of dimensions
      of the Bigarray.

      Each element [Genarray.get b i] is initialized to the result of [f i].
      In other words, [Genarray.init kind layout dimensions f] tabulates
      the results of [f] applied to the indexes of a new big array whose
      layout is described by [kind], [layout] and [dimensions].  The index
      array [i] may be shared and mutated between calls to f.

      For instance, [Genarray.init int c_layout [|2; 1; 3|]
      (Array.fold_left (+) 0)] returns a fresh big array of integers, in C
      layout, having three dimensions (2, 1, 3, respectively), with the
      element values 0, 1, 2, 1, 2, 3.

      [Genarray.init] raises [Invalid_argument] if the number of dimensions
      is not in the range 0 to 16 inclusive, or if one of the dimensions
      is negative. *)
end

module Array0 : sig
  (* ... *)

  val init: ('a, 'b) kind -> 'c layout -> 'a -> ('a, 'b, 'c) t
  (** [Array0.init kind layout v] behaves like [Array0.create kind layout]
     except that the element is additionally initialized to the value [v]. *)
end

module Array1 : sig
  (* ... *)

  val init: ('a, 'b) kind -> 'c layout -> int -> (int -> 'a) ->
            ('a, 'b, 'c) t
  (** [Array1.init kind layout dim f] returns a new Bigarray [b]
     of one dimension, whose size is [dim].  [kind] and [layout]
     determine the array element kind and the array layout
     as described for {!Genarray.create}.

     Each element [Array1.get b i] of the array is initialized to the
     result of [f i].

     In other words, [Array1.init kind layout dimensions f] tabulates
     the results of [f] applied to the indexes of a new big array whose
     layout is described by [kind], [layout] and [dim]. *)
end

module Array2 : sig
  (* ... *)

  val init: ('a, 'b) kind ->  'c layout -> int -> int ->
            (int -> int -> 'a) -> ('a, 'b, 'c) t
  (** [Array2.init kind layout dim1 dim2 f] returns a new Bigarray [b]
     of two dimensions, whose size is [dim2] in the first dimension
     and [dim2] in the second dimension.  [kind] and [layout]
     determine the array element kind and the array layout
     as described for {!Bigarray.Genarray.create}.

     Each element [Array2.get b i j] of the array is initialized to
     the result of [f i j].

     In other words, [Array2.init kind layout dim1 dim2 f] tabulates
     the results of [f] applied to the indexes of a new Bigarray whose
     layout is described by [kind], [layout], [dim1] and [dim2]. *)
end

module Array3 : sig
  (* ... *)

  val init: ('a, 'b) kind ->  'c layout -> int -> int -> int ->
            (int -> int -> int -> 'a) -> ('a, 'b, 'c) t
  (** [Array3.init kind layout dim1 dim2 dim3 f] returns a new Bigarray [b]
     of three dimensions, whose size is [dim1] in the first dimension,
     [dim2] in the second dimension, and [dim3] in the third.
     [kind] and [layout] determine the array element kind and the array
     layout as described for {!Bigarray.Genarray.create}.

     Each element [Array3.get b i j k] of the array is initialized to
     the result of [f i j k].

     In other words, [Array3.init kind layout dim1 dim2 dim3 f] tabulates
     the results of [f] applied to the indexes of a new Bigarray whose
     layout is described by [kind], [layout], [dim1], [dim2] and [dim3]. *)
end
```

### CODE REVIEW

File: stdlib/bigarray.ml [hunk 1]
```diff
@@ -99,6 +99,36 @@ module Genarray = struct
      = "caml_ba_get_generic"
   external set: ('a, 'b, 'c) t -> int array -> 'a -> unit
      = "caml_ba_set_generic"
+
+  let rec cloop arr idx f col max =
+    if col = pred (Array.length idx) then
```
Comment (gasche) [inline stdlib/bigarray.ml]: The code would be nicer with

```ocaml
if col = Array.length idx then
  set arr idx (f idx)
else ...
```

**Review by gasche — COMMENTED** [review]
In the `GenArray` case I find the use of `int array` a bit frustrating as we have to specify that mutating the index is not good (it would be nicer to have immutable arrays or possibly just just `(int -> int)`), but this is consistent with the rest of the module so I agree it is the better choice.

File: stdlib/bigarray.ml (hunk 1 above)
Comment (yallop) [reply stdlib/bigarray.ml]: That's a good idea; thanks (af87f4e).

**Comment by gasche** [comment]
This looks like a sensible feature to have to me (especially given that bigarrays currently do not have `make` functions to initialize with a constant, if I understand correctly), and I would be in favor of inclusion.

I think the initialization order for fortran-layout bigarrays (with the first coordinate varying faster than the other ones) slightly surprising, naively I would have expected the same order as for C-layout bigarrays (with the last coordinate varying faster). Is there a reason for the implemented behavior?

**Comment by yallop** [comment]
> I think the initialization order for fortran-layout bigarrays […] slightly surprising[…]. Is there a reason for the implemented behavior?

The reason is to initialize elements in the same order they're laid out in memory.  For example, suppose that you initialize a C-layout and a Fortran-layout array by returning `1`, `2`, `3`, … from successive calls to the function passed to `init`:

```ocaml
let countup () = let r = ref 0 in fun _ -> (incr r; !r)
let carray = Genarray.init int c_layout       [|2;3;4|] (countup ());;
let farray = Genarray.init int fortran_layout [|4;3;2|] (countup ());;
```

Then, given a function which views arrays through a flat C layout:
```ocaml
let flatten arr = let len = Array.fold_left ( * ) 1 (Genarray.dims arr) in
                  Array1.change_layout (reshape_1 arr len) c_layout
```

your arrays will enjoy the pleasant property that
```ocaml
flatten carray = flatten farray
```

There's something aesthetically appealing about this, but it's practical, too: initializing Fortran-layout arrays in this order is faster (sometimes 2× or more, and often significantly).

**Comment by xavierleroy** [comment]
> There's something aesthetically appealing about this, but it's practical, too: initializing Fortran-layout arrays in this order is faster (sometimes 2× or more, and often significantly).

Locality, locality, locality!  For best performance you want the minor dimension to vary faster than the major dimensions.  For the C layout the minor dimension is the last dimension, and for the Fortran layout it's the first.

**Review by gasche — APPROVED** [review]
Thanks for the clarification. I had no idea the Fortran layout worked like that, I thought it was mostly about 1-based indexing.

I'm out of questions or suggestions for this PR and I think it is clearly a good function to have, so I'm approving. This should be good to go if the CI is also happy.

File: stdlib/bigarray.mli [hunk 2]
```diff
@@ -298,6 +298,32 @@ module Genarray :
...
+
+      For instance, [Genarray.init int c_layout [|2; 1; 3|]
+      (Array.fold_left (+) 0)] returns a fresh big array of integers, in C
+      layout, having three dimensions (2, 1, 3, respectively), with the
+      element values 0, 1, 2, 1, 2, 3.
+
+      [Genarray.init] raises [Invalid_argument] if the number of dimensions
+      is not in the range 0 to 16 inclusive, or if one of the dimensions
+      is negative. *)
+
```
Comment (nojb) [inline stdlib/bigarray.mli:328]: "since" annotation missing (here and below).

Comment (yallop) [reply stdlib/bigarray.mli:328]: Thanks (29e78e7).

**Review by xavierleroy — APPROVED** [review]
`Array{1,2,3}.init` are very nice, `Genarray.init` is as good as it can be (but nobody will want to use it), and `Array0.init` is slightly pedantic.  

Speaking of pedantry, in the documentation comments I would have used "indices" as the plural of "index", but that's probably because I'm a Latin pedant.

OK for merging as is.

**Comment by yallop** [comment]
> Speaking of pedantry, in the documentation comments I would have used "indices" as the plural of "index", but that's probably because I'm a Latin pedant.

I decided to fix this (12d0a359) since "indices" is used elsewhere in the file.

**Comment by dra27** [comment]
Would you like this squashed or merged, @yallop?

> Speaking of pedantry, in the documentation comments I would have used "indices" as the plural of "index", but that's probably because I'm a Latin pedant.

❤

**Comment by yallop** [comment]
> Would you like this squashed or merged, @yallop?

Thanks for asking.  I'm happy with either.

**Pull request merged**